NGitLab/Impl/VariableClient.cs

```diff
@@ -18,9 +20,20 @@ protected VariableClient(API api, string urlPrefix)
 
     public Variable this[string key] => _api.Get().To<Variable>(_urlPrefix + "/variables/" + key);
 
+    [EditorBrowsable(EditorBrowsableState.Never)]
     public Variable Create(VariableCreate model) => _api.Post().With(model).To<Variable>(_urlPrefix + "/variables");
 
-    public Variable Update(string key, VariableUpdate model) => _api.Put().With(model).To<Variable>(_urlPrefix + "/variables/" + key);
+    public Variable Create(Variable model) => _api.Post().With(model).To<Variable>(_urlPrefix + "/variables");
 
-    public void Delete(string key) => _api.Delete().Execute(_urlPrefix + "/variables/" + key);
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public Variable Update(string key, VariableUpdate model) => _api.Put().With(model).To<Variable>($"{_urlPrefix}/variables/{key}");
+
+    public Variable Update(string key, Variable model) => Update(key, model, null);
+
+    public Variable Update(string key, Variable model, string environmentScope) => _api.Put().With(model).To<Variable>(string.IsNullOrWhiteSpace(environmentScope) ? $"{_urlPrefix}/variables/{key}" : $"{_urlPrefix}/variables/{key}?filter[environment_scope]={Uri.EscapeDataString(environmentScope!)}");
```

Comment:
As the environment scope is already a property in the Variable class, I would just take the property directly to make the condition, instead of having another param in the methods.

It might change a bit the tests, but it would feels probably more natural when using the client ^^

Reply:
The idea behind this signature was that this way you can actually change the environment scope of an existing variable. If I rely solely on the property EnvironmentScope I would need to delete the existing and create a new Variable.

@Toa741 What are your thoughts given that additional piece of information?

Reply:
Ok, I see the reason you did it as such. I'll check again if there would be another or if we go with that

Reply:
I checked a bit the API:
- We need the variable Key and potentially the scope to construct the URL in any case, with its corresponding variable values
- As such, I would maybe specify the environment scope as the second param (to keep the URL item along)?
- For the Variable object to pass:
  - If we follow what you did in the test (`new Variable { Value = "My value edited", Protected = false, EnvironmentScope = newScope}`), for me it would make sense maybe to pass a separate object (like an Updated VariableUpdate class), as such we make sure that we can't just pass an existing variable that we modified directly. And it will contains only the modified properties (others will be let as null). It would behave like a 'patch'.
  - Or, we can let the possiblity to pass Variable class directly in it, though, with the key/scope to pass as well as parameters, it might be messy to understand.
- Also, I would make the param names clearer:
  - key => currentVariableKey
  - environmentScope => currentVariableEnvironmentScope
  - variable => updatedVariable (if it's a Variable class) or variableUpdate (if it's a VariableUpdate class)

@louis-z, I wonder about your input on this? The VariableUpdate sounds like there would be less misunderstanding for me.

Reply:
I think we should keep using `VariableCreate` and `VariableUpdate`, as this will allow us to clearly define which properties can/must be set for creation and updating. Yes, this means adding the missing properties there too, not just in the `Variable` class, but how often will this change afterwards? Not often, I think.

Reply:
- Keeping the key and the environmentScope in the signature make sense, as they are required (and used in the URL)
- The change I would make it to put the  environmentScope as the second argument to keep the key and the scope close as they are part of the "query", not the update in itself

```csharp
Variable Update(string key, VariableUpdate model);
Variable Update(string key, string environmentScope, VariableUpdate model);
```

For the `VariableCreate/Update`, it could make a difference in term of visibility and to avoid missuses.
- It prevents using an existing `Variable` directly (less confusion for the update case with the current key/scope to use)
- It enforce the fact that the returned value should be used instead of a modified version of the `Variable` object
- It avoids using partial `Variable` declaration as in the test (that's what made me questioned about keeping  `VariableCreate/Update`)
- Not all properties seem used actually between a create and update (at least on v15 while playing with the API directly, I can't update the key. Didn't tested on v16, but could be quickly done by writing a test :))